sphinx/environment/__init__.py

```diff
@@ -358,17 +357,16 @@ def merge_info_from(self, docnames, other, app):
         app.emit('env-merge-info', self, docnames, other)
 
     def path2doc(self, filename):
-        # type: (unicode) -> unicode
+        # type: (unicode) -> Optional[unicode]
         """Return the docname for the filename if the file is document.
 
         *filename* should be absolute or relative to the source directory.
         """
         if filename.startswith(self.srcdir):
-            filename = filename[len(self.srcdir) + 1:]
+            filename = os.path.relpath(filename, self.srcdir)
         for suffix in self.config.source_suffix:
-            if fnmatch.fnmatch(filename, '*' + suffix):
-                return filename[:-len(suffix)]
-        # the file does not have docname
+            if filename.endswith(suffix):
+                return filename.rstrip(suffix)
```

Comment:
I noticed this is not good. `str.rstrip()` is not alternative of `filename[:-len(suffix]`.
```
>>> "rrrr.rst".rstrip(".rst")
''
```